csharp/src/Ice/Endpoint.cs

```diff
@@ -131,5 +133,124 @@ public abstract class Endpoint : IEquatable<Endpoint>
         // Return a server side transceiver for this endpoint, or null if a transceiver can only be created by an
         // acceptor.
         public abstract ITransceiver? GetTransceiver();
+
+        // Creates an endpoint from a string.
+        internal static Endpoint Parse(string endpointString, Communicator communicator, bool oaEndpoint)
+        {
+            string[]? args = IceUtilInternal.StringUtil.SplitString(endpointString, " \t\r\n");
+            if (args == null)
+            {
+                throw new FormatException($"mismatched quote in endpoint `{endpointString}'");
+            }
+
+            if (args.Length == 0)
+            {
+                throw new FormatException("no non-whitespace character in endpoint string");
+            }
+
+            string transport = args[0];
+            if (transport == "default")
+            {
+                transport = communicator.DefaultTransport;
+            }
+
+            var options = new Dictionary<string, string?>();
+
+            // Parse args into options (and skip transport at args[0])
+            for (int n = 1; n < args.Length; ++n)
+            {
+                // Any option with < 2 characters or that does not start with - is illegal
+                string option = args[n];
+                if (option.Length < 2 || option[0] != '-')
+                {
+                    throw new FormatException($"invalid option `{option}' in endpoint `{endpointString}'");
+                }
+
+                // Extract the argument given to the current option, if any
+                string? argument = null;
+                if (n + 1 < args.Length && args[n + 1][0] != '-')
+                {
+                    argument = args[++n];
+                }
+
+                try
+                {
+                    options.Add(option, argument);
+                }
+                catch (ArgumentException)
+                {
+                    throw new FormatException($"duplicate option `{option}' in endpoint `{endpointString}'");
+                }
+            }
+
+            if (communicator.FindEndpointFactory(transport) is IEndpointFactory factory)
+            {
+                Endpoint endpoint = factory.Create(endpointString, options, oaEndpoint);
+                if (options.Count > 0)
+                {
+                    throw new FormatException(
+                        $"unrecognized option(s) `{ToString(options)}' in endpoint `{endpointString}'");
+                }
+                return endpoint;
+            }
+
+            // If the stringified endpoint is opaque, create an unknown endpoint, then see whether the type matches one
+            // of the known endpoints.
+            if (!oaEndpoint && transport == "opaque")
```

Comment:
Why check for oaEndpoint here? Although it's unlikely that a user would ever specify an OA endpoint as an "opaque" endpoint I don't see why we should restrict parsing opaque endpoints only for certain contexts (or we'd have to document in which context it's allowed).

Reply:
We are parsing an endpoint string here, and `oaEndpoint` means it's an endpoint-string given for OA configuration. You should not specify "opaque" as your transport when you configure your OA endpoints. Documenting this restriction is of course fine.

`opaque` is useful for proxies you receive over the wire and want to print even though the local Ice runtime doesn't known all the transports of this proxy. It should not be used to configure OA endpoints, even in a round-about manner where the transport is actually known to the Ice runtime.

Reply:
Isn't it the case that I can set `Ice.Default.Locator` with a stringified opaque endpoint and it's not going to be rejected while it will be rejected for the OA endpoint property? 

To me it's not very consistent to reject one but not the other... and yes in both cases it's a twisted way to setup endpoints 😄. Who knows... perhaps it's an activation service that receives the endpoints as proxies, writes a configuration file and start an Ice process with this configuration file...! 

I understand it's useful to receive proxies with unknown endpoints and be able to stringify them... but here we are talking about parsing the string back into an endpoint.

Reply:
We have this `oaEndpoint` bool parameter because  proxy and OA endpoints are not handled the same way. See the columns "client semantics" and "server semantics" on https://doc.zeroc.com/ice/3.7/client-side-features/proxies/proxy-and-endpoint-syntax.

So not accepting the `opaque` transport for OA endpoints is far from the only difference between proxy and OA endpoints--they are already quite different. And I don't think you can make a reasonable argument for `opaque` as a reasonable OA endpoint transport.